### PR TITLE
feat(autofix): Add controllable insights chain

### PR DESCRIPTION
**Pull request by roaga**
When an certain update request is received, we jump back to the selected point in the steps/insights chain, delete any steps, insights, and memory after that, add the user feedback to the memory, and continue analyzing.

This allows the user full control over the model's reasoning, and let's them rethink and guide the model at any point. This also allows restarting from errored/no-root cause states, and jumping back to previous steps.

### CODE REVIEW

File: src/seer/automation/autofix/event_manager.py [hunk 1]
```diff
@@ -65,6 +65,8 @@ def restart_step(self, step: Step):
         with self.state.update() as cur:
             cur_step = cur.find_or_add(step)
             cur_step.status = AutofixStatus.PROCESSING
+            cur_step.progress = []
+            cur_step.completedMessage = None  # type: ignore[assignment]
```
Comment (roaga) [inline src/seer/automation/autofix/event_manager.py:69]: For some reason mypy complained about this assignment even though `completedMessage` is defined as `Optional`. I couldn't figure it out, so I put this ignore, but if you see a valid issue, do let me know.

File: src/seer/automation/autofix/steps/steps.py [hunk 2]
```diff
@@ -45,7 +56,11 @@ def _invoke(self, **kwargs: Any) -> Any:
 
     def _pre_invoke(self) -> bool:
         # Don't run the step instance if it's already been run
-        return make_done_signal(self.request.step_id) not in self.context.state.get().signals
+        if make_done_signal(self.request.step_id) not in self.context.state.get().signals:
+            self.thread = threading.Thread(target=self._check_for_kill)
```
Comment (roaga) [inline src/seer/automation/autofix/steps/steps.py:60]: Please double-check all the threading stuff, I don't want to break anything

File: src/seer/automation/autofix/tasks.py [hunk 3]
```diff
@@ -311,10 +315,112 @@ def receive_user_message(request: AutofixUpdateRequest):
...
+        if count > 5:
+            break
+
+    event_manager = AutofixEventManager(state)
+    context = AutofixContext(
+        state=state, sentry_client=get_sentry_client(), event_manager=event_manager
+    )
+    step_to_restart = cast(DefaultStep, state.get().steps[-1])
+
+    is_coding_step = any(
```
Comment (jennmueng) [inline src/seer/automation/autofix/tasks.py:355]: you can check the step `key` for this, there is a `find_step` method: https://github.com/getsentry/timeseries-analysis-service/blob/db1652d16da8e3906df2a1e60c0b61558b4c869f/src/seer/automation/autofix/models.py#L329-L335

File: src/seer/automation/autofix/tasks.py [hunk 4]
```diff
@@ -311,10 +315,112 @@ def receive_user_message(request: AutofixUpdateRequest):
...
+                            + num_tool_calls
+                            + 1
+                        ]
+                    )
+                else:
+                    new_memory = new_memory[:-1]
+            memory = new_memory
+            break
+    if not step_to_restart.insights:
+        memory = memory[:1]
```
Comment (jennmueng) [inline src/seer/automation/autofix/tasks.py:385]: looks like this can be its own method as well, this way we can have a specific unit test for it

**Review by jennmueng — COMMENTED** [review]
a lot of the logic with steps should be its own method and well tested imo, right now the logic looks kinda fragile

File: src/seer/automation/autofix/steps/steps.py [hunk 5]
```diff
@@ -91,11 +106,25 @@ def _get_extra_invoke_kwargs(self) -> dict[str, Any]:
...
+    def _check_for_kill(self):
+        while True:
+            if self.thread_kill:
+                return
+            kill_signal = make_kill_signal()
+            if kill_signal in self.context.state.get().signals:
+                with self.context.state.update() as cur:
+                    cur.signals.remove(kill_signal)
+                self.thread_kill = True
+                os._exit(1)
```
[the remark this answers is not in the document]
Comment (roaga) [reply src/seer/automation/autofix/steps/steps.py:125]: Idk what you mean by automatically restart. It kills this step, and then we start a new job for the step after in `tasks.py`

[the remark this answers is not in the document]
Comment (jennmueng) [reply src/seer/automation/autofix/steps/steps.py]: So each celery worker starts N number of processes right, does this just permanently kill a given process or does it restart

[the remark this answers is not in the document]
Comment (roaga) [reply src/seer/automation/autofix/steps/steps.py]: I guess I don't fully understand how our celery workers work.

Does starting a new step = starting a new worker? And is os.kill the right choice or something else?

[the remark this answers is not in the document]
Comment (jennmueng) [reply src/seer/automation/autofix/steps/steps.py]: so there is 1 celery worker, that starts N number of processes. Starting a new step schedules a task that is picked up by the worker and assigned to a process, that process runs the task, and once it's done becomes "free" again

[the remark this answers is not in the document]
Comment (roaga) [reply src/seer/automation/autofix/steps/steps.py]: Ok, so is what I'm doing right now killing the process without restarting it? How can I restart the process, or free it up without killing it?

**Review by jennmueng — APPROVED** [review]

**Comment by roaga** [comment]
👋 Hi there! Here is a root cause analysis of the issue above automatically generated by Autofix 🤖

This analysis was triggered by admin@sentry.io.

# Multiple Calls to `bootup` Leading to `RuntimeError`

## Description
The `RuntimeError` occurs because the `bootup` function is called multiple times across the codebase. Each call to `bootup` results in a call to `initialize_database`, which attempts to register the `SQLAlchemy` instance on the Flask app. Since the `SQLAlchemy` instance is already registered, this leads to the `RuntimeError`. Ensure that `bootup` is called only once during the application lifecycle to prevent multiple registrations of the `SQLAlchemy` instance.

## Reproduction Steps

1. Ensure that the `bootup` function is called multiple times in the application lifecycle, in different parts of the codebase such as `src/celery_app/app.py`, `src/seer/app.py`, and `src/seer/grpc.py`. 
2. Verify that each call to `bootup` results in a call to `initialize_database`, leading to multiple registrations of the `SQLAlchemy` instance on the same Flask app.

## Relevant Code Context

### `bootup` Function in `src/seer/bootup.py`
The `bootup` function initializes various components, including the database, logging, and models. It's critical that this function is only called once to prevent duplicate registrations.

**File:** src/seer/bootup.py
**Repository:** getsentry/seer
```
@inject
def bootup(
    *, start_model_loading: bool, integrations: list[Integration], config: AppConfig = injected
):
    initialize_sentry_sdk(integrations)
    with sentry_sdk.metrics.timing(key="seer_bootup_time"):
        initialize_logs(["seer.", "celery_app."])
        config.do_validation()
        initialize_database()  # <-- SUSPECT LINE
        initialize_models(start_model_loading)

```

### `initialize_database` Function in `src/seer/db.py`
This function configures the database connection for a Flask app using SQLAlchemy. Multiple initializations can lead to runtime errors due to re-registration attempts.

**File:** src/seer/db.py
**Repository:** getsentry/seer
```
@inject
def initialize_database(
    config: AppConfig = injected,
    app: Flask = injected,
):
    app.config["SQLALCHEMY_DATABASE_URI"] = config.DATABASE_URL
    app.config["SQLALCHEMY_ENGINE_OPTIONS"] = {"connect_args": {"prepare_threshold": None}}

    db.init_app(app)  # <-- SUSPECT LINE
    migrate.init_app(app, db)

```

## More Info

If you have any questions or feedback for the Sentry team about this fix, please email [autofix@sentry.io](mailto:autofix@sentry.io) with the Run ID (see below).

### 🤓 Stats for the nerds:
Run ID: **1**
Prompt tokens: **69137**
Completion tokens: **5368**
Total tokens: **74505**

**Comment by roaga** [comment]
👋 Hi there! Here is a root cause analysis of the issue above automatically generated by Autofix 🤖

This analysis was triggered by admin@sentry.io.

# Multiple Registrations of SQLAlchemy Instance and AttributeError with PromiseProxy

## Description
We have identified two potential root causes for the issues observed:

1. **Multiple Registrations of SQLAlchemy Instance**:
   - The `RuntimeError` is caused by multiple registrations of the `SQLAlchemy` instance on the Flask app.
   - This occurs when `db.init_app(app)` is called more than once in the application lifecycle within the `initialize_database` function.
   - To resolve this, ensure that `initialize_database` is called only once during the application lifecycle.

2. **AttributeError with PromiseProxy**:
   - The `AttributeError` involving `PromiseProxy` is likely due to an issue with an external library.
   - The error occurs when a `PromiseProxy` object without the expected `__thing` attribute is passed to a function wrapped by the `inject` decorator.
   - Since `PromiseProxy` is not defined within the codebase, further investigation into the external library is required.

## Reproduction Steps
### Reproduction Steps

1. **Multiple Registrations of SQLAlchemy Instance**:
   - Attempt to call `initialize_database()` multiple times within the same application lifecycle, ensuring that the `app` variable is a `Flask` instance and `config` is an `AppConfig` instance.

2. **AttributeError with PromiseProxy**:
   - Invoke a function decorated with `@inject` that relies on dependency injection, ensuring that the `argspec` includes a `PromiseProxy` object, which lacks the `__thing` attribute.

## Relevant Code Context

### initialize_database function in `src/seer/db.py`
This snippet defines how the database is initialized within the application lifecycle. The function `initialize_database` shows where the dual SQLAlchemy initialization can occur.

**File:** src/seer/db.py
**Repository:** getsentry/seer
```
@inject
def initialize_database(
    config: AppConfig = injected,
    app: Flask = injected,
):
    app.config["SQLALCHEMY_DATABASE_URI"] = config.DATABASE_URL
    app.config["SQLALCHEMY_ENGINE_OPTIONS"] = {"connect_args": {"prepare_threshold": None}}

    db.init_app(app)  # <-- SUSPECT LINE
    migrate.init_app(app, db)
```

### wrapper function in `src/seer/dependency_injection.py`
This function manages the dependency injection process, and is suspected to cause issues when unexpected objects such as 'PromiseProxy' are encountered.

**File:** src/seer/dependency_injection.py
**Repository:** getsentry/seer
```
def inject(c: _A) -> _A:
    original_type = c
    if inspect.isclass(c):
        c = c.__init__

    argspec = inspect.getfullargspec(c)

    @functools.wraps(c)  # type: ignore
    def wrapper(*args, **kwargs):
        new_kwds = {**kwargs}

        if argspec.defaults:
            offset = len(argspec.args) - len(argspec.defaults)
            for i, d in enumerate(argspec.defaults):
                arg_idx = offset + i
                arg_name = argspec.args[arg_idx]

                if d is injected and len(args) <= arg_idx and arg_name not in new_kwds:
                    try:
                        resolved = resolve(argspec.annotations[arg_name])
                    except KeyError:
                        raise AssertionError(
                            f"Cannot inject argument {arg_name} as it lacks annotations"
                        )

                    new_kwds[arg_name] = resolved

        if argspec.kwonlydefaults:
            for k, v in argspec.kwonlydefaults.items():
                if v is injected and k not in new_kwds:
                    try:
                        new_kwds[k] = resolve(argspec.annotations[k])
                    except KeyError:
                        raise AssertionError(f"Cannot inject argument {k} as it lacks annotations")

        return c(*args, **new_kwds)  # type: ignore  # <-- SUSPECT LINE
```

## More Info

If you have any questions or feedback for the Sentry team about this fix, please email [autofix@sentry.io](mailto:autofix@sentry.io) with the Run ID (see below).

### 🤓 Stats for the nerds:
Run ID: **2**
Prompt tokens: **71605**
Completion tokens: **4444**
Total tokens: **76049**

**Pull request merged**